### PR TITLE
Expose API for sysroot search & compile flags

### PR DESCRIPTION
**Pull request by Kha**
Since `leanc` is not meant to be used for generic C compilation anymore, instead expose its flags in a convenient manner for Lake. Also expose a function inspired by Lake's code for locating the sysroot, which can also be useful for other tools (/cc @insightmind).

### CODE REVIEW

**Comment by tydeu** [comment]
@Kha Neato! A few minor comments:

1. I would still really like to be able access the flags alone directly (without the include / library path). That way I can still use the richer `LeanInstall` of Lake to fill in the paths. The combined function is quite nice for a lot of simpler use cases (so I would keep it as well).

2. I would name `findLeanSysroot?` just `findSysroot?` since it is already in the `Lean` namespace.

File: src/Lean/Compiler/FFI.lean [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+/-- Return C compiler flags for including Lean's headers. -/
+def getCFlags (leanSysroot : FilePath) : Array String :=
+  #["-I", (leanSysroot / "include").toString] ++ (getLeancExtraFlags ()).trim.splitOn
+
+@[extern "lean_get_linker_flags"]
+private constant getBuiltinLinkerFlags (linkStatic : Bool) : String
+
+/-- Return linker flags for linking against Lean's libraries. -/
+def getLinkerFlags (leanSysroot : FilePath) (linkStatic := true) (gmp := "-lgmp") : Array String :=
+  #["-L", (leanSysroot / "lib" / "lean").toString, gmp] ++ (getBuiltinLinkerFlags linkStatic).trim.splitOn
```
Comment (tydeu) [inline src/Lean/Compiler/FFI.lean:23]: My specific suggestion for the flag file is:

```suggestion
@[extern "lean_get_leanc_extra_flags"]
private constant getLeancExtraFlags : Unit → String

/-- Return C compiler flags for compiling C files outputted by Lean. -/
def extraCFlags : Array String :=
  getLeancExtraFlags () |>.trim.splitOn

/-- Return C compiler flags (with include path) for compiling C files outputted by Lean. -/
def cFlags (leanSysroot : FilePath) : Array String :=
  #["-I", (leanSysroot / "include").toString] ++ extraCFlags

@[extern "lean_get_linker_flags"]
private constant getBuiltinLinkerFlags (linkStatic : Bool) : String

/-- Return linker flags for linking against Lean's libraries. -/
def extraLinkerFlags (linkStatic := true) : Array String :=
  getBuiltinLinkerFlags linkStatic |>.trim.splitOn

/-- Return linker flags (with library paths) for linking against Lean's libraries. -/
def linkerFlags (leanSysroot : FilePath) (linkStatic := true) (gmp := "-lgmp") : Array String :=
  #["-L", (leanSysroot / "lib" / "lean").toString, gmp] ++ extraLinkerFlags
```

**Comment by tydeu** [comment]
@Kha Also, a point of confusion for me: why are the `leanc` internal C/linker flags not part of the exposed flags?

**Comment by Kha** [comment]
They wouldn't be all that internal then, would they!

**Comment by Kha** [comment]
>     1. That way I can still use the richer `LeanInstall` of Lake to fill in the paths.

Please specify? Why does it matter whether Lean or Lake assembles the path if the outcome is the same?

**Comment by Kha** [comment]
>     3\. Also, style-wise, why are the flag functions call `get*` (which is usually reserved for impure/lazy/moadic code)

I'm not aware of such a convention? In contrast, I believe that most functions should contain a verb.

**Comment by tydeu** [comment]
> > ```
> > 1. That way I can still use the richer `LeanInstall` of Lake to fill in the paths.
> > ```
> 
> Please specify? Why does it matter whether Lean or Lake assembles the path if the outcome is the same?

Technically, the paths of `LeanInstall` are separate so the `oleanDir` doesn't necessarily have to be in the standard location. Though, admittedly, that is unlikely to occur.

**Comment by tydeu** [comment]
> > ```
> > 3\. Also, style-wise, why are the flag functions call `get*` (which is usually reserved for impure/lazy/moadic code)
> > ```
> 
> I'm not aware of such a convention? In contrast, I believe that most functions should contain a verb.

I was basing this comment mostly on how the `Lean.version` definitions were patterned. I took the arguments more as templating for the flags (i.e., like a constructor which usually does not have verb).  I just generally associate the verb `get` in functional programming with state (after all that is what plain `get` deals with) not pure functions. But that's just my personal preference, feel free to ignore my comment.

**Pull request merged**